File: dotnet/source/amp.messaging/MessageRegistration.cs
using System;
using System.Collections.Generic;
using cmf.bus;
using Common.Logging;

namespace amp.messaging
{
    public class MessageRegistration : IRegistration
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(MessageRegistration));

        private readonly IMessageHandler _handler;
        private readonly Predicate<Envelope> _filterPredicate;
        private readonly Func<Envelope, MessageContext> _envelopeOpener;
        private IDictionary<string, string> _info;


        public Predicate<Envelope> Filter
        {
            get { return _filterPredicate; }
        }

        public IDictionary<string, string> Info
        {
            get { return _info; }
        }

        public object Handle(Envelope env)
        {
            Log.Debug("Enter MessageRegistration # handle( Envelope env )");

            try {
                // use the provided envelope opener to get a context
                MessageContext context = _envelopeOpener(env);

                if (null != context)
                {
                    _handler.Handle(context.Message, context.Envelope.Headers);
                }
                else
                {
                    // if our opener returned null, that signifies that it declined
                    // to open it, and that we should quietly drop this envelope
                    // So, I'll return null here to make this explicit, even though
                    // just doing nothing here would lead to the 'return null' that
                    // is just a few lines below this.
                    return null;
                }
            }
            catch (MessageException ex) {
                const string message = "Failed to process an incoming message envelope.";
                Log.Error(message, ex);
                throw new Exception(message, ex);
            }

            return null;
        }

        public object HandleFailed(Envelope env, Exception ex)
        {
            Log.Debug("Enter HandleFailed");

            // either succeed & return the result or fail & return null
            try
            {
                object handleFailed = _handler.HandleFailed(env, ex);
                Log.Debug("Leave HandleFailed");
                return handleFailed;
            }
            catch (Exception failedToFail)
            {
                Log.Warn("Caught an exception attempting to handle the failed event", failedToFail);
                return null;
            }
        }

        public MessageRegistration(Func<Envelope, MessageContext> envelopeOpener, IMessageHandler handler)
            :this(envelopeOpener, handler, env => true)
        {
        }

        public MessageRegistration(
            Func<Envelope, MessageContext> envelopeOpener,
            IMessageHandler handler,
            Predicate<Envelope> filterPredicate)
        {
            _envelopeOpener = envelopeOpener;
            _handler = handler;
            _filterPredicate = filterPredicate;

            string handledType = _handler.Topic;

            _info = new Dictionary<string, string>
            {
                {EnvelopeHeaderConstants.MESSAGE_TOPIC, handledType},
                {EnvelopeHeaderConstants.MESSAGE_TYPE, handledType}
            };
        }
    }
}

File: java/amp-messaging/src/main/java/amp/messaging/MessageRegistration.java
package amp.messaging;


import java.util.HashMap;
import java.util.Map;

import cmf.bus.Envelope;
import amp.messaging.EnvelopeHeaderConstants;
import cmf.bus.IEnvelopeFilterPredicate;
import cmf.bus.IRegistration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * Created with IntelliJ IDEA.
 * User: jar349
 * Date: 5/1/13
 */
public class MessageRegistration implements IRegistration {

    private static final Logger LOG = LoggerFactory.getLogger(MessageRegistration.class);

    private IMessageProcessor _processor;
    private IMessageHandler _handler;
    private IEnvelopeFilterPredicate _predicate;
    private Map<String, String> _regInfo;


    @Override
    public Map<String, String> getRegistrationInfo() { return _regInfo; }

    @Override
    public IEnvelopeFilterPredicate getFilterPredicate() { return _predicate; }


    public MessageRegistration(
            IMessageProcessor processor,
            IMessageHandler handler,
            IEnvelopeFilterPredicate predicate) {

        _processor = processor;
        _handler = handler;
        _predicate = predicate;

        _regInfo = new HashMap<String, String>();
        _regInfo.put(EnvelopeHeaderConstants.MESSAGE_TOPIC, _handler.getMessageType().getCanonicalName());
        _regInfo.put(EnvelopeHeaderConstants.MESSAGE_TYPE, _handler.getMessageType().getCanonicalName());
    }


    @Override
    public Object handle(final Envelope env) throws Exception {

        LOG.debug("Enter MessageRegistration # handle( Envelope env )");

        try {
            // create a context to send through the processors
            final MessageContext ctx = new MessageContext(MessageContext.Directions.In, env);

            _processor.processMessage(ctx, new IContinuationCallback() {

                @Override
                public void continueProcessing() throws MessageException {
                    _handler.handle(ctx.getMessage(), env.getHeaders());
                }
            });

        }
        catch (MessageException ex) {
            String message = "Failed to process an incoming envelope.";
            LOG.error(message, ex);
            throw new Exception(message, ex);
        }

        return null;
    }

    @Override
    public Object handleFailed(Envelope env, Exception ex) throws Exception {
        return null;
    }
}
